### PR TITLE
[New feature] Add HySparse MQA/MLA block-sparse attention TileLang ops

### PR DESCRIPTION
**Pull request by ForFishes**
### PR Category
Operator Mechanism

### PR Types
New features

### Description

Add TileLang implementations of the HySparse (arXiv 2602.03560) block attention pipeline for the MQA/MLA setting (single shared K/V head), under `src/paddlefleet/tilelang_ops/hysparse/`:

- **block-score attention** (fwd + bwd): full attention that additionally emits per-(query, key-block) max logits for downstream TopK block selection. Causal + document masking via **document-relative** block coordinates, with causal/document early-exit to skip fully-masked blocks.
- **block-sparse gather attention** (fwd + bwd): per-token block-sparse attention over TopK-selected blocks; block indices shared across heads and one shared K/V head so a single gathered block feeds every query head.
- **pipeline**: block-score scoring -> group-wise (cross-head) max aggregation -> TopK block selection -> block-sparse attention.
- **reference.py**: naive Paddle fp32 reference used as ground truth.
- **benchmark.py**: standalone performance harness.

Supports `head_dim=576` (MLA) and `batch>1`. The backward tiles the query `M` dim and sub-tiles the key `N` dim to fit shared memory at large head dims (Blackwell ~227KB/block).

**Tested**: 25 single-card precision tests in `tests/single_card_tests/custom_ops/test_hysparse_block_attn.py`, all validated against the naive fp32 reference (causal / document / `block_B in {32,64,128}` / `H=1` / deep-doc / empty rows / block_N sub-tiling invariance / batch=2 / packed multi-doc bit-exact / D=576 MLA fwd+bwd / topk>num_blocks shape). Tolerances: out < 5e-2, lse < 1e-3, block-score < 1e-3, grad < 8e-2 (bf16 matmul + fp32 accumulation).

**Review follow-ups addressed**:
- `select_topk_blocks` always returns `[B, S, topk]` (surplus slots padded with -1 when `topk > num_blocks`) and validates `topk > 0`.
- `ref_block_sparse_attn_mqa` empty-row LSE stays `-inf` (matches kernel/docstring).
- Lightweight host-side shape checks added to the forward interfaces.
- Removed the test's `sys.path.insert(src)` hack (relies on the installed package).

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/tilelang_ops/hysparse/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+    valid = _valid_block_mask(valid_range, num_blocks, block_B)  # [B,S,nb]
+    neg = paddle.full_like(agg, -1.0)
+    agg = paddle.where(valid, agg, neg)  # invalid blocks pushed to the bottom
+
+    k = min(topk, num_blocks)
+    top_val, top_idx = paddle.topk(agg, k=k, axis=-1)  # [B, S, k]
+    # slots that landed on an invalid block (negative score) -> -1
+    top_idx = paddle.where(top_val >= 0, top_idx, paddle.full_like(top_idx, -1))
+    return top_idx.astype("int32").contiguous()
```
Comment (risemeup1111) [inline src/paddlefleet/tilelang_ops/hysparse/pipeline.py]: P2 **优先级：P2**

**问题：** 当 `topk > num_blocks` 时，这里把 `k` 截断为 `num_blocks` 后直接返回，`select_topk_blocks` / `hysparse_forward_mqa` 的 `indices` 形状会变成 `[B, S, num_blocks]`，与文档承诺的 `[B, S, topk]` 不一致。短序列或较大的默认 `topk` 下，调用方如果按请求的 `topk` 维度分配/检查张量，会遇到形状不稳定的问题。

**影响：** 这是边界输入下的 API 契约问题；即使语义上已经选择了全部 block，也应该把多余 slot 保留为 `-1`，和“slots beyond the number of valid blocks are -1”的说明一致。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

建议保持 `paddle.topk` 的 `k <= num_blocks`，但返回前补齐到请求的 `topk`：

```suggestion
    k = min(topk, num_blocks)
    top_val, top_idx = paddle.topk(agg, k=k, axis=-1)  # [B, S, k]
    # slots that landed on an invalid block (negative score) -> -1
    top_idx = paddle.where(top_val >= 0, top_idx, paddle.full_like(top_idx, -1))
    if k < topk:
        pad = paddle.full([b, s, topk - k], -1, dtype=top_idx.dtype)
        top_idx = paddle.concat([top_idx, pad], axis=-1)
    return top_idx.astype("int32").contiguous()
```

**Review by risemeup1111 — COMMENTED** [review]
已完成首轮检查，未发现需要阻塞合入的 P0/P1 问题。一个边界形状问题已放在行级评论里，请作者回复或修正。

- 非行级：PR 描述没有按仓库模板填写 `PR Category` / `PR Types` / `是否引起精度变化`，当前 `Check PR Description` 因此失败。P3 **优先级：P3** 请按模板补充，例如 `PR Category: Operator Mechanism`（或更合适的分类）、`PR Types: New features`、`是否引起精度变化: 否/是（按实际情况）`。**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by ForFishes** [comment]
感谢 review！已在 5af7b01 修复：

1. **[P2/Bug] `select_topk_blocks` 形状契约**（@risemeup1111 @PaddlePaddle-bot pipeline.py:93）：`topk > num_blocks` 时现在始终返回 `[B, S, topk]`，多余 slot 用 `-1` 补齐（gather kernel 与参考实现本就忽略 `-1`）；并在入口校验 `topk > 0`。
2. **[Copilot] `ref_block_sparse_attn_mqa` 空行 LSE**（reference.py）：改为在 `-inf` masked logits 上直接取 LSE，空行保持 `-inf`，与 kernel/docstring 一致；softmax 的 NaN 空行仍由 `row_has_key` 置零。
3. **[Copilot] 接口 shape 校验**：block-score / block-sparse 前向接口新增轻量 host 端 shape 断言（不触发 device 同步）。
4. **[PaddlePaddle-bot 🟡] 测试 `sys.path.insert`**：已移除，改为依赖已安装包（与 CI 构建的 per-PR whl 一致，`Build Fleet whl` 已通过说明子包已正确打包）。
5. 新增 `test_topk_exceeds_num_blocks` 覆盖补齐后的索引宽度；本地 25 个单测全过。

另外已按模板补全 PR 描述（PR Category / PR Types / 是否引起精度变化）。

**Review by risemeup1111 — APPROVED** [review]
已复查 5af7b01 的修改并回应作者的更新说明：此前 `topk > num_blocks` 导致 `indices` 宽度不稳定的问题已修复，返回宽度现在保持 `[B, S, topk]` 并用 `-1` 补齐，新增的回归测试也覆盖了该场景。PR 描述模板项已补全，最新的描述检查已有成功记录；旧的失败记录看起来是更新前的结果。

本轮未发现新的需要阻塞合入的问题。仍在运行中的测试请以最终 CI 结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by sneaxiy — APPROVED** [review]
LGTM for HySparse operators.
